src/OrchardCore.Modules/OrchardCore.AuditTrail/Services/AuditTrailManager.cs

```diff
@@ -73,27 +74,42 @@ public async Task AddAuditTrailEventAsync<TAuditTrailEventProvider>(AuditTrailCo
                     auditTrailContext.EventFilterKey,
                     auditTrailContext.EventFilterData);
 
-                _auditTrailEventHandlers.Invoke((handler, context)
+                await _auditTrailEventHandlers.InvokeAsync((handler, context)
                     => handler.CreateAsync(context), auditTrailCreateContext, Logger);
-                
+
                 var auditTrailEvent = new AuditTrailEvent
                 {
                     Id = _iidGenerator.GenerateUniqueId(),
                     Category = eventDescriptor.CategoryDescriptor.Category,
                     EventName = auditTrailCreateContext.EventName,
                     FullEventName = eventDescriptor.FullEventName,
-                    UserName = !string.IsNullOrEmpty(auditTrailCreateContext.UserName) ? auditTrailCreateContext.UserName : T["[empty]"],
+                    UserName = !string.IsNullOrEmpty(auditTrailCreateContext.UserName)
+                        ? auditTrailCreateContext.UserName
+                        : T["[empty]"],
                     CreatedUtc = auditTrailCreateContext.CreatedUtc ?? _clock.UtcNow,
                     Comment = auditTrailCreateContext.Comment.NewlinesToHtml(),
                     EventFilterData = auditTrailCreateContext.EventFilterData,
                     EventFilterKey = auditTrailCreateContext.EventFilterKey,
-                    ClientIpAddress = string.IsNullOrEmpty(auditTrailCreateContext.ClientIpAddress) ?
-                        await GetClientAddressAsync() : auditTrailCreateContext.ClientIpAddress
+                    ClientIpAddress = string.IsNullOrEmpty(auditTrailCreateContext.ClientIpAddress)
+                        ? await GetClientAddressAsync()
+                        : auditTrailCreateContext.ClientIpAddress
                 };
 
                 eventDescriptor.BuildAuditTrailEvent(auditTrailEvent, auditTrailCreateContext.EventData);
 
+                await _auditTrailEventHandlers.InvokeAsync((handler, context, @event)
+                    => handler.AlterAsync(context, @event), auditTrailCreateContext, auditTrailEvent, Logger);
+
                 _session.Save(auditTrailEvent);
+
+                // This forces the session to not skip the FlushAsync right after. Without that if two threads add documents
+                // at the same time (as can happen with AuditTrailEvents) the second will not be added into the internal
+                // document registry (IdentityMap._documents) corrupting the state and causing an "Incorrect attempt to
+                // update an object that doesn't exist..." exception.
+                typeof(Session)
+                    .GetField("_flushing", BindingFlags.Instance | BindingFlags.NonPublic)!
+                    .SetValue((Session)_session, false);
+                await _session.FlushAsync();
```

Comment:
> if two threads add documents at the same time (as can happen with AuditTrailEvents)

This should not happen in the context of a given shell scope, so in the context of a given scoped session, async calls are awaited to return to the caller in a non blocking way, but all of them are completed sequentially in the execution flow of a given shell scope, otherwise this would mean that some async calls are not awaited.

Reply:
Did you check this @domonkosgabor?